### ISSUE TITLE
Graph.add_node method mutates attr_dict parameter

### ISSUE THREAD

**Issue by thegreathippo** [networkx/networkx#2105]
```
>>> import networkx
>>> g = networkx.Graph()
>>> my_dict = {"x": 0, "y": 0}
>>> g.add_node("node", my_dict, y=1)
>>> print(my_dict)
{'y': 1, 'x': 0}
>>> my_dict["x"] = 1
>>> print(g.node["node"])
{'y': 1, 'x': 1}
>>> my_other_dict = {"x": 5}
>>> g.add_node("node", my_other_dict)
>>> print(my_dict)
{'y': 1, 'x': 5}
>>> g.add_node("other node", my_dict)
>>> my_dict["x"] = 10
>>> print(g.node["node"])
{'y': 1, 'x': 10}
>>> print(g.node["other node"])
{'y': 1, 'x': 10}
```
Is this working as-intended? It seems like it could create some very subtle, very hard-to-track bugs -- such as someone using the same dictionary to initialize multiple nodes without realizing that by doing so, they've caused all those nodes to share the same dictionary -- or someone naively using an object's `__dict__` to generate node attributes, not realizing that by doing so, they'll modify that object's attributes every time they modify the node attributes (or vice versa).

(I suppose there are some potential edge cases where this might be beneficial? Maybe you _want_ certain nodes to share a dictionary, or maybe you _want_ to be able to modify node attributes from outside the graph object? But both those things seem like a bad idea to me, and something you could probably handle by _directly_ assigning a dictionary rather than doing it through the add_node method)

I also noticed that add_edge avoids part of this issue by creating a new dict ('datadict') and updating it with attr_dict -- though it still mutates attr_dict with the **kwargs passed to it (so you'll always end up with a fresh edge attribute dictionary, but you'll still mutate the dictionary you passed to the method to create that dictionary). If this wasn't done, then it would be very easy to have multiple edges share the same dictionary (which would potentially play havoc when it came time to start relying on the 'weight' edge attribute!).

If this isn't intentional, it seems easy to fix (just use a copy of attr_dict rather than attr_dict); if it _is_ intentional, there might be some benefit to documenting the difference between how add_node assigns node attributes and how add_edge assigns edge attributes?

(Apologies in advance if this issue has already been raised / addressed!)

**Comment by hagberg** [comment]
I agree this is somewhat surprising behavior.  I suggest that we do not mutate the user input attribute dictionary.  I'd prefer to not make the `add_node` and `add_edge` methods more complicated to so let's see if we come up with a good  solution.  @dschult what do you think?

**Comment by thegreathippo** [comment]
I took a shot at fixing this myself (keep in mind, it's my first time trying to contribute to a github project, so I might have screwed something up big time); I put in the pull request [here](https://github.com/networkx/networkx/pull/2107).

**Comment by hagberg** [comment]
Fixed in #2107 #2132
